0023-keymanager-secret-sharing.md

```diff
@@ -0,0 +1,671 @@
+  announces the new committee and begins collecting registrations for the
+  next handoff.
+
+- The old committee deletes obsolete full shares.
+
+- The committee starts serving requests.
+
+### Identification
+
+```go
+// NodeToID assigns a unique ID to a node for use in the CHURP protocol.
+func NodeToID(nodeID signature.PublicKey) []byte {
+  id := nodeID
+  return id[:]
+}
+```
+
+### Configuration
+
+```go
+// SchnorrPrime is a Schnorr prime p, where p = qr + 1, for some prime q.
+type SchnorrPrime struct {
+  // P is the prime number p.
+  P []byte `json:"p"`
+  // Q is the prime factor of p-1.
+  Q []byte `json:"q"`
+  // R is the reminder of p-1 when divided by q.
+  R uint32 `json:"r"`
+}
+
+// SchnorrGroup is a large prime-order subgroup of the multiplicative group
+// of integers modulo Schnorr prime p.
+type SchnorrGroup struct {
+  // Modulus is a Schnorr prime.
+  Modulus SchnorrPrime `json:"modulus"`
+  // Generator is a generator of the group.
+  Generator []byte `json:"generator"`
+}
+
+// Config contains the CHURP configuration.
+type Config struct {
+  // Group is a Schnorr group.
+  Group SchnorrGroup `json:"group"`
+
+  // Threshold is the minimum number of distinct shares required to reconstruct a secret.
+  Threshold uint8 `json:"threshold"`
+
+  // HandoffInterval is the time interval in epochs between handoffs.
+  //
+  // Zero value disables handoffs.
+  HandoffInterval beacon.EpochTime `json:"handoff_interval,omitempty"`
+
+  // BlameAssignment is true iff the responses should be checked
+  // for corrupted shares.
+  BlameAssignment bool `json:"blame_assignment,omitempty"`
+}
+
+// PolicySGX is a CHURP access control policy.
+type PolicySGX struct {
+  // Serial is the monotonically increasing policy serial number.
+  Serial uint32 `json:"serial"`
+
+  // ID is the runtime ID that this policy is valid for.
+  RuntimeID common.Namespace `json:"runtime_id"`
+
+  // ID is the CHURP instance ID that this policy is valid for.
+  ChurpID uint8 `json:"churp_id"`
+
+  // MayQuery is the map of runtime IDs to the vector of enclave IDs that
+  // may query private key material.
+  MayQuery map[common.Namespace][]sgx.EnclaveIdentity `json:"may_query,omitempty"`
+
+  // MayShare is the vector of enclave IDs from which a share can be obtained
+  // during handouts.
+  MayShare []sgx.EnclaveIdentity `json:"may_share,omitempty"`
+
+  // MayJoin is the vector of enclave IDs that may join the new committee
+  // during handoffs.
+  MayJoin []sgx.EnclaveIdentity `json:"may_join,omitempty"`
+}
+
+// SignedPolicySGX is a signed SGX CHURP access control policy.
+type SignedPolicySGX struct {
+  Policy PolicySGX `json:"policy"`
+
+  Signatures []signature.Signature `json:"signatures"`
+}
+```
+
+### Consensus transactions
+
+```go
+var (
+  // MethodChurpCreate is the method name for creating a new CHURP instance.
+  MethodChurpCreate = transaction.NewMethodName(
+    ModuleName, "Churp/Create", Config{},
```

Comment:
Maybe we should discuss how to handle this additional method namespacing.

Reply:
`app.Method`?

Reply:
So then the complete method name would be `keymanager.app.Method`? That would work.